token/transfer-hook-interface/src/onchain.rs

```diff
@@ -49,3 +52,40 @@ pub fn execute<'a>(
     )?;
     invoke(&cpi_instruction, &cpi_account_infos)
 }
+
+/// Helper to add accounts required for the transfer-hook program on-chain, looking
+/// through the additional account infos to add the proper accounts
+pub fn add_cpi_accounts_for_execute<'a>(
+    cpi_instruction: &mut Instruction,
+    cpi_account_infos: &mut Vec<AccountInfo<'a>>,
+    mint_pubkey: &Pubkey,
+    program_id: &Pubkey,
+    additional_accounts: &[AccountInfo<'a>],
+) -> ProgramResult {
+    let validation_pubkey = get_extra_account_metas_address(mint_pubkey, program_id);
+    let validation_info = additional_accounts
+        .iter()
+        .find(|&x| *x.key == validation_pubkey)
+        .ok_or(TransferHookError::IncorrectAccount)?;
+
+    let program_info = additional_accounts
+        .iter()
+        .find(|&x| x.key == program_id)
+        .ok_or(TransferHookError::IncorrectAccount)?;
+
+    ExtraAccountMetas::add_to_cpi_instruction::<instruction::ExecuteInstruction>(
+        cpi_instruction,
+        cpi_account_infos,
+        &validation_info.try_borrow_data()?,
+        additional_accounts,
+    )?;
+    cpi_account_infos.push(validation_info.clone());
+    cpi_account_infos.push(program_info.clone());
```

Comment:
The ordering of accounts here jumps out at me a bit. In the `instruction.rs` of `transfer-hook-interface` you have:
```text
///   0. `[]` Source account
///   1. `[]` Token mint
///   2. `[]` Destination account
///   3. `[]` Source account's owner/delegate
///   4. `[]` Validation account
///   5..5+M `[]` `M` additional accounts, written in validation account data
```

Reply:
What's cool is that it doesn't matter! The helpers pull them out in the right order, and in a few other places it's also using this ordering. It's true that it might look unclear though, so I'll fix up the ordering everywhere

Reply:
Or just outline with a comment that the helper muscles that! Good stuff!